### PR TITLE
Add `foldSubmap` and `submap`

### PR DESCRIPTION
**Pull request by hdgarrood**
Addresses #71 and #113.

/cc @mtolly can you check whether this works for you? I don't think we need both an ascending and descending version because you can use [Data.Monoid.Dual](https://pursuit.purescript.org/packages/purescript-monoid/3.0.0/docs/Data.Monoid.Dual) if you want it in the other direction.

### CODE REVIEW

**Comment by hdgarrood** [comment]
Also @mtolly I think the newtype I mentioned in #113 does exist: it's `App` from `functors`, although it doesn't yet have that `Monoid` instance. See https://github.com/purescript/purescript-functors/issues/12

**Comment by mtolly** [comment]
Yep, looks great to me. Thanks for the help!

File: src/Data/Map.purs [hunk 1]
```diff
@@ -254,6 +256,109 @@ findMin Leaf = Nothing
...
+-- |  == ["one", "two"]
+-- |
+-- | foldSubmap Nothing (Just 2) (\_ v -> [v])
+-- |  (fromFoldable [Tuple 0 "zero", Tuple 1 "one", Tuple 2 "two", Tuple 3 "three"])
+-- |  == ["zero", "one", "two"]
+-- | ```
+foldSubmap :: forall k v m. Ord k => Monoid m => Maybe k -> Maybe k -> (k -> v -> m) -> Map k v -> m
+foldSubmap kmin kmax f =
+  let
+    tooSmall =
```
Comment (paf31) [inline src/Data/Map.purs]: Thought: you could use `any` here to make this a little shorter at the expense of a function call and a dictionary lookup.

**Comment by paf31** [comment]
👍 Looks great!

**Pull request merged**